### PR TITLE
make: Optimize for Westmere by default on x86_64

### PR DESCRIPTION
**Pull request by brson**
## What have you changed? (mandatory)

This adds a new file, opts.mk, included in Makefile, that sets values of RUSTFLAGS, CXXFLAGS, and CFLAGS to optimize for a specific baseline architecture. For x86_64 on Linux and OS X this is Westmere, a 2010-era CPU, per https://github.com/tikv/tikv/issues/4843. For other platforms, nothing has changed.

These baseline optimizations are enabled by default, but can be controlled by the `OPTIMIZE` environment variable as described in the comments. The makefiles don't appear to support cross-compilation today, but this file is set up to support it in the future by obeying a `TARGET` variable if it is set.

I still need to benchmark this patch, but am submitting it for review of the general approach.

Fixes https://github.com/tikv/tikv/issues/4843

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)
- Misc (other changes)

## How has this PR been tested? (mandatory)

Watching strace to see that the correct flags are passed to compilers.

Still need to test that the code contains new optimizations and has performance wins.

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

**Comment by brson** [comment]
cc on CI doesn't like `-march=westmere` ...

**Comment by breeswish** [comment]
Do we need to change some dependencies' build script like rust-rocksdb for this to take effect?

**Comment by Connor1996** [comment]
The `Westmere` arch supports `AVX`, right? If using that arch as the baseline, the binary can not run on KVM virtual processor due to the absence of `AVX` support. We already encountered that issue before, See https://github.com/tikv/tikv/issues/4999

**Comment by siddontang** [comment]
Great work @brson

**Comment by siddontang** [comment]
can you help us to check whether it can work or not on KVM?

/cc @lichunzhu

**Comment by breeswish** [comment]
According to [wikipedia](https://en.wikipedia.org/wiki/SSE4) and [gcc documentation](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html) looks like you should target at `nehalem` instead of `westmere`, which is the first microarchitecture that includes SSE 4.2. Also according to the gcc documentation both `westmere` and `nehalem` does not include AVX instructions which should be able to reduce @Connor1996 's concern.

**Comment by Connor1996** [comment]
@breeswish You're right. The Chinese version of Wikipedia may be wrong and it misleads me. It says `westmere` supports `AVX`, while the English version does not.

File: opts.mk [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+# -march=westmere, but _tune_ the CPU for -mtune=generic (-mcpu is like -march +
+# -mtune). We are explicitly setting -march and -mtune for gcc.
+#
+# rustc only has -Ctarget-cpu, which is used to annotate all functions with an
+# LLVM "target-cpu" attribute, and also passed to the gcc linker plugin as
+# -mcpu, so here we are telling gcc/clang to use e.g. westmere instructions
+# while using a "generic" x86_64 tuning model, but _seemingly_ telling LLVM to
+# use westmere instructions with a westmere tuning model.
+#
+# The impact of this in not clear - it may be that LLVM doesn't distinguish
```
Comment (overvenus) [inline opts.mk]: What about replacing `-Ctarget-cpu` with `-Ctarget-feature=some backend features`?

https://github.com/rust-lang/rfcs/blob/master/text/2045-target-feature.md

**Comment by lichunzhu** [comment]
> can you help us to check whether it can work or not on KVM?
> 
> /cc @lichunzhu

@siddontang I have tried to compile `tikv` using `make build` from `brson:march` and the compiled `tikv` works and passes all tests on my KVM. In the next step, I need to use binary `tikv` compiled by newer CPUs to check whether it really works. So where can I get that?

**Comment by Hoverbear** [comment]
@brson Can you help @lichunzhu ? :)

**Comment by brson** [comment]
Here are the benches before and after: https://gist.github.com/7a245b46a62ea742a81293644412c0f8

As is often the case, it's pretty hard to interpret. Some wins some loses. Highlights:

- writebatch benchmarks improved by up to 27%
- codec benchmarks improved up to 18%
- number::benches::bench_encode_u64_le_byteorder regressed by 40%!
- storage fixture benches had -15% to 18% changes
- index_scan_index/RocksDB/ regressed by 57%
- criterion benches seem to have regressed more than improved

**Comment by brson** [comment]
Here are the benchmark results: https://gist.github.com/7d6d2fe00933ace95bf1d415d43c6a09

Here's the script I used to collect them: https://gist.github.com/3de50bd40d8ac1fd2cd4f0aef4d50ac6

**Comment by brson** [comment]
It's hard to interpret the benchmarks, and these are run on AWS so I don't trust the non-criterion benches. Some improvements some regressions, so I don't consider this a performance win.

I think this should be merged so that there is some framework in place for knowing the target architecture.

**Comment by brson** [comment]
I pushed a new commit that corrects some comments.

**Comment by brson** [comment]
PTAL @nrc @breeswish @overvenus

**Comment by breeswish** [comment]
I will try personally to see whether or not all components including 3rd-party libraries are covered by these flags.. If so, I will give my approval.

**Comment by brson** [comment]
Closing this so I can open other PRs.

**Comment by Hoverbear** [comment]
@breeswish PTAL

**Comment by brson** [comment]
Reopening. Previous measurements indicated this PR performed worse than master. I've merged with master to retest again.

@zhouqiang-cl how do I have this PR automatically benchmarked by CI?

**Comment by brson** [comment]
Closing this PR again to open others.

**Pull request closed without merging**